test/types/community/collection/recursive-types.test-d.ts

```diff
@@ -1,23 +1,70 @@
-import { expectError } from 'tsd';
+import { expectAssignable, expectError, expectNotAssignable, expectNotType } from 'tsd';
 
-import type { Collection } from '../../../../src';
+import type { Collection, Filter, UpdateFilter } from '../../../../src';
 
 /**
  * mutually recursive types are not supported and will not get type safety
  */
-interface A {
-  b: B;
+interface Author {
+  name: string;
+  favoritePublication: Book;
 }
 
-interface B {
-  a: A;
+interface Book {
+  title: string;
+  author: Author;
 }
 
-declare const mutuallyRecursive: Collection<A>;
-//@ts-expect-error
-mutuallyRecursive.find({});
-mutuallyRecursive.find({
-  b: {}
+expectAssignable<Filter<Author>>({
+  favoritePublication: {
+    title: 'book title',
+    author: {
+      name: 'author name'
+    }
+  }
+});
+expectNotType<UpdateFilter<Author>>({
+  $set: {
+    favoritePublication: {
+      title: 'a title',
+      published: new Date(),
+      author: {
+        name: 23
+      }
+    }
+  }
+});
+
+// Extremely deep type checking for recursive schemas
+expectNotAssignable<Filter<Author>>({
+  'favoritePublication.author.favoritePublication.author.favoritePublication.author.favoritePublication.title': 23
+});
+expectAssignable<Filter<Author>>({
+  'favoritePublication.author.favoritePublication.author.favoritePublication.author.favoritePublication.author.favoritePublication.title':
+    'good soup'
+});
+expectNotAssignable<Filter<Author>>({
+  'favoritePublication.author.favoritePublication.author.favoritePublication.author.name': 23
+});
+
+// Beyond the depth of 10, `extends Document` permits anything (number for name is permitted)
+expectAssignable<Filter<Author>>({
+  'favoritePublication.author.favoritePublication.author.favoritePublication.author.favoritePublication.author.favoritePublication.author.name': 23
+});
+
+// Update filter has similar depth limit
+expectAssignable<UpdateFilter<Author>>({
+  $set: {
+    'favoritePublication.author.favoritePublication.author.favoritePublication.author.favoritePublication.author.name':
```

Comment:
wouldn't 9 be not type checked? also, if update is "similar", then we should use the same test cases (-1 length in each) and be clear about how they are similar, i.e., it's one less because there is an extra level of inference

Reply:
I changed the prop to bestBook because I was honestly losing count of the depth due to long strings, but now we have a test for each level on both filter types.